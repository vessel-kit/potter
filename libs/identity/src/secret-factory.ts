import { Ed25519Secret, ISecret, Secp256k1Secret } from './secret';
import { KeyKind } from './key-kind';
import * as sha256 from '@stablelib/sha256';
import { InvalidSecretKindError } from "./invalid-secret-kind.error";

const encoder = new TextEncoder();

export class SecretFactory {
  fromSeed(kind: KeyKind.secp256k1, seed: Uint8Array | string): Secp256k1Secret;
  fromSeed(kind: KeyKind.ed25519, seed: Uint8Array | string): Ed25519Secret;
  fromSeed(kind: KeyKind, seed: Uint8Array | string): ISecret {
    const bytes = typeof seed === 'string' ? encoder.encode(seed) : seed;
    const material = sha256.hash(bytes);
    switch (kind) {
      case KeyKind.ed25519:
        return new Ed25519Secret(material);
      case KeyKind.secp256k1:
        return new Secp256k1Secret(material);
      default:
        throw new InvalidSecretKindError(kind);
    }
  }
}
