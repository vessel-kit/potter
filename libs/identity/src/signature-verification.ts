import * as elliptic from 'elliptic';
import * as ed25519 from '@stablelib/ed25519';
import BN from 'bn.js';

const secp256k1Context = new elliptic.ec('secp256k1');

export class SignatureVerification {
  secp256k1(publicKey: Uint8Array, message: Uint8Array, signature: Uint8Array): boolean {
    const keyPair = secp256k1Context.keyFromPublic(publicKey);
    const r = new BN(signature.slice(0, 32));
    const s = new BN(signature.slice(32, 64));
    return keyPair.verify(message, { r, s });
  }

  ed25519(publicKey: Uint8Array, message: Uint8Array, signature: Uint8Array) {
    return ed25519.verify(publicKey, message, signature);
  }
}
