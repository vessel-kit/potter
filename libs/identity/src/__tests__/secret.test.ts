import { SecretFactory } from '../secret-factory';
import { KeyKind } from '../key-kind';
import * as sha256 from '@stablelib/sha256';
import { SignatureVerification } from '../signature-verification';

const signatureVerification = new SignatureVerification();
const secretFactory = new SecretFactory();
const seed = 'foo';

describe('Secp256k1Secret', () => {
  const secret = secretFactory.fromSeed(KeyKind.secp256k1, seed);
  const digest = sha256.hash(new Uint8Array([1, 2, 3]));
  test('sign', async () => {
    const signature = await secret.sign(digest);
    expect(signature).toMatchSnapshot();
    const pub = await secret.publicKey();
    expect(signatureVerification.secp256k1(pub.material, digest, signature)).toBeTruthy();
  });
});

describe('Ed25519Secret', () => {
  const secret = secretFactory.fromSeed(KeyKind.ed25519, seed);
  const digest = sha256.hash(new Uint8Array([1, 2, 3]));
  test('sign', async () => {
    const signature = await secret.sign(digest);
    expect(signature).toMatchSnapshot();
    const pub = await secret.publicKey();
    expect(signatureVerification.ed25519(pub.material, digest, signature)).toBeTruthy();
  });
});
