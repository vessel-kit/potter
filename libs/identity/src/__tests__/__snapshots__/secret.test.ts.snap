// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ed25519Secret sign 1`] = `
Uint8Array [
  127,
  100,
  118,
  137,
  179,
  103,
  204,
  69,
  217,
  59,
  156,
  253,
  18,
  29,
  109,
  41,
  226,
  154,
  110,
  199,
  132,
  77,
  52,
  118,
  161,
  126,
  170,
  172,
  225,
  37,
  142,
  22,
  38,
  9,
  123,
  107,
  58,
  118,
  169,
  79,
  130,
  22,
  177,
  134,
  110,
  172,
  244,
  28,
  217,
  28,
  230,
  245,
  71,
  195,
  182,
  77,
  63,
  100,
  51,
  155,
  211,
  16,
  53,
  1,
]
`;

exports[`Secp256k1Secret sign 1`] = `
Uint8Array [
  7,
  195,
  62,
  218,
  77,
  136,
  58,
  222,
  170,
  141,
  141,
  90,
  163,
  83,
  234,
  146,
  237,
  174,
  66,
  230,
  14,
  214,
  143,
  85,
  201,
  165,
  242,
  91,
  62,
  214,
  154,
  44,
  48,
  233,
  70,
  174,
  234,
  124,
  168,
  222,
  205,
  6,
  158,
  30,
  232,
  135,
  232,
  100,
  252,
  86,
  210,
  201,
  36,
  128,
  29,
  166,
  73,
  75,
  36,
  207,
  87,
  158,
  133,
  120,
]
`;
